torch/utils/checkpoint.py

```diff
@@ -364,78 +732,65 @@ def _checkpoint_without_reentrant(function, preserve_rng_state=True, *args, **kw
-
-                with torch.enable_grad(), \
-                     torch.cuda.amp.autocast(**gpu_autocast_kwargs), \
-                     torch.cpu.amp.autocast(**cpu_autocast_kwargs), \
-                     torch.autograd.graph.saved_tensors_hooks(inner_pack, inner_unpack):
-                    _unused = function(*args, **kwargs)
-
-        if x not in storage:
-            raise RuntimeError(
-                "Attempt to retrieve a tensor saved by autograd multiple times without checkpoint"
-                " recomputation being triggered in between, this is not currently supported. Please"
-                " open an issue with details on your use case so that we can prioritize adding this."
-            )
-
-        return storage[x]
+    # From checkpoint_wrapper.
+    # We should modify to handle non-tensor, kwargs
+    flat_args, kwarg_keys = _pack_kwargs(*args, **kwargs)
 
-    with torch.autograd.graph.saved_tensors_hooks(pack, unpack):
-        output = function(*args, **kwargs)
+    def new_fn(*inputs):
+        # This function should be called immediately by checkpoint_impl
+        unpacked_args, unpacked_kwargs = _unpack_kwargs(
+            inputs, kwarg_keys
+        )
+        out = fn(*unpacked_args, **unpacked_kwargs)
         if torch.cuda._initialized and preserve_rng_state and not had_cuda_in_fwd:
             # Cuda was not initialized before running the forward, so we didn't
             # stash the CUDA state.
             raise RuntimeError(
                 "PyTorch's CUDA state was initialized in the forward pass "
                 "of a Checkpoint, which is not allowed. Please open an issue "
                 "if you need this feature.")
+        return out
 
-    return output
+    def recompute_fn(*inputs):
+        # This will be called later during recomputation. This wrapping enables
+        # the necessary global state to be captured.
+        unpacked_args, unpacked_kwargs = _unpack_kwargs(
+            inputs, kwarg_keys
+        )
+
+        rng_devices = []
+        if preserve_rng_state and had_cuda_in_fwd:
+            rng_devices = fwd_gpu_devices
+        with torch.random.fork_rng(devices=rng_devices, enabled=preserve_rng_state):
+            if preserve_rng_state:
+                torch.set_rng_state(fwd_cpu_state)
+                if had_cuda_in_fwd:
+                    set_device_states(fwd_gpu_devices, fwd_gpu_states)
+
+            with torch.cuda.amp.autocast(**gpu_autocast_kwargs), \
+                 torch.cpu.amp.autocast(**cpu_autocast_kwargs):
+                fn(*unpacked_args, **unpacked_kwargs)
+
+    return _checkpoint_impl(new_fn, recompute_fn, *flat_args)
+
+def _checkpoint_impl(fn, recompute_fn, *args):
+    global curr_is_checkpoint, curr_is_recompute
+    new_frame = _CheckpointFrame(recompute_fn)
+
+    # See [ Implementation details of Rule 2 ]
+    if curr_is_checkpoint:
+        args, grad_fn = _applyAutogradFunctionToSaveInputs(*args)
```

Comment:
My experience from the Module full backward hooks is that, while this works, it is a bit disruptive as people do tend to do inplace.
If we don't have a choice, this is ok. But if we could avoid it, we should

Reply:
Good point, will look at alternatives.

Reply:
I don't think inplace works well with checkpoint actually